src/common/config_validator.cpp

```diff
@@ -0,0 +1,142 @@
+                   "Model file does not exist: " + modelFile);
+  }
+}
+
+void ConfigValidator::validateOptionsParallelData() const {
+  auto trainSets = get<std::vector<std::string>>("train-sets");
+  UTIL_THROW_IF2(trainSets.empty(),
+                 "No train sets given in config file or on command line");
+
+  auto vocabs = get<std::vector<std::string>>("vocabs");
+  UTIL_THROW_IF2(!vocabs.empty() && vocabs.size() != trainSets.size(),
+                 "There should be as many vocabularies as training sets");
+}
+
+void ConfigValidator::validateOptionsScoring() const {
+  boost::filesystem::path modelPath(get<std::string>("model"));
+
+  UTIL_THROW_IF2(!boost::filesystem::exists(modelPath),
+                 "Model file does not exist: " + modelPath.string());
+  UTIL_THROW_IF2(get<std::vector<std::string>>("vocabs").empty(),
+                 "Scoring, but vocabularies are not given!");
+}
+
+void ConfigValidator::validateOptionsTraining() const {
+  auto trainSets = get<std::vector<std::string>>("train-sets");
+
+  UTIL_THROW_IF2(
+      has("embedding-vectors")
+          && get<std::vector<std::string>>("embedding-vectors").size()
+                 != trainSets.size(),
+      "There should be as many embedding vector files as training sets");
+
+  boost::filesystem::path modelPath(get<std::string>("model"));
+
+  auto modelDir = modelPath.parent_path();
+  if(modelDir.empty())
+    modelDir = boost::filesystem::current_path();
+
+  UTIL_THROW_IF2(
+      !modelDir.empty() && !boost::filesystem::is_directory(modelDir),
+      "Model directory does not exist");
+
+  UTIL_THROW_IF2(!modelDir.empty()
+                     && !(boost::filesystem::status(modelDir).permissions()
+                          & boost::filesystem::owner_write),
+                 "No write permission in model directory");
+
+  UTIL_THROW_IF2(has("valid-sets")
+                     && get<std::vector<std::string>>("valid-sets").size()
+                            != trainSets.size(),
+                 "There should be as many validation sets as training sets");
+
+  // validations for learning rate decaying
+  UTIL_THROW_IF2(get<double>("lr-decay") > 1.0,
+                 "Learning rate decay factor greater than 1.0 is unusual");
+
+  auto strategy = get<std::string>("lr-decay-strategy");
+
+  UTIL_THROW_IF2(
+      (strategy == "epoch+batches" || strategy == "epoch+stalled")
+          && get<std::vector<size_t>>("lr-decay-start").size() != 2,
+      "Decay strategies 'epoch+batches' and 'epoch+stalled' require two "
+      "values specified with --lr-decay-start option");
+  UTIL_THROW_IF2(
+      (strategy == "epoch" || strategy == "batches" || strategy == "stalled")
+          && get<std::vector<size_t>>("lr-decay-start").size() != 1,
+      "Single decay strategies require only one value specified with "
+      "--lr-decay-start option");
+}
+
+void ConfigValidator::validateDevices(cli::mode mode) const {
```

Comment:
Does it make sense to keep a sepaerate validate funciton? Why can the validate function not just call the actual function that interprets this, and let it fail?

Reply:
It doesn't. I moved it here from the old ConfigParser. 
I've made it private and now it's called only from validateOptions().